### DISCUSSION TITLE
Upgrading to Phalcon 3 Centos8 broke the existing site

### DISCUSSION THREAD

**Discussion by jwang628** [phalcon/cphalcon#15317] [General]
Hi All,

the most related discussion of my issue is here:
https://forum.phalcon.io/discussion/13099/upgrading-to-phalcon-3-broke-the-existing-site 

However, i failed to locate Module.php

Please shed some light.

Thanks a lot in advance

**Comment by Jeckerson** [comment]
Hello

From what version of Phalcon do you migrating?

**Reply by jwang628** [reply]
Version which worked (Windows 10):

php -v
PHP 5.6.32 (cli) (built: Oct 25 2017 16:00:49) 
Copyright (c) 1997-2016 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2016 Zend Technologies



Web framework delivered as a C-extension for PHP
phalcon => enabled
Author => Phalcon Team and contributors
Version => 3.2.4
Build Date => Oct 23 2017 11:07:17
Powered by Zephir => Version 0.10.4-11e39849b0

One Centos8 (which has issue):
/opt/remi/php72/root/usr/bin/php -v 
PHP 7.2.34 (cli) (built: Feb  3 2021 08:32:06) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies

Web framework delivered as a C-extension for PHP
phalcon => enabled
Author => Phalcon Team and contributors
Version => 3.4.5
Build Date => Oct 31 2019 08:39:29
Powered by Zephir => Version 0.10.16


Not sure why it breaks on Centos8

**Comment by jwang628** [comment]
the line in index.php which breaks (error 500):
$application->handle()->getContent()

the following works though:
/opt/remi/php72/root/usr/bin/php index.php

**Reply by Jeckerson** [reply]
Try to upgrade to PHP `7.3`. Also check webserver logs.

**Reply by jwang628** [reply]
thanks a lot.  but there is no error logged.

**Reply by jwang628** [reply]
php73 also has the same error:

HTTP error 500: Internal Server Error                                                                                                                                                  
                                                                                                                                                                                       
   ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────   
                                                                                                                                                                                       
   An error occurred on the server while fetching the document you requested. However, the server did not send back any explanation of what happened, so it is unknown what went       
   wrong. Please contact the web server administrator about this, if you believe that this error should not occur since it is not a nice behaviour from the web server at all and      
   indicates that there is some much deeper problem with the web server software.                                                                                                      
                                                                                                                                                                                       
   URI: 
   

the following page (via apache) showed:
<?php
echo "hello world";
// Show all information, defaults to INFO_ALL
phpinfo();

?>

   PHP logo                                                                                                                                                                            
                                                                                                                                                                                       
   PHP Version 7.3.27                                                                                                                                                                  
                                                                                                                                                                                       
   System                                  Linux testvm2 4.18.0-147.8.1.el8_1.x86_64 #1 SMP Thu Apr 9 13:49:54 UTC 2020 x86_64                                                         
   Build Date                              Feb 2 2021 10:32:50                                                                                                                         
   Build System                            Red Hat Enterprise Linux release 8.3 (Ootpa)                                                                                                
   Build Provider                          Remi's RPM repository <https://rpms.remirepo.net/>                                                                                          
   Server API                              FPM/FastCGI                                                                                                                                 
   Virtual Directory Support               disabled                                                                                                                                    
   Configuration File (php.ini) Path       /etc/opt/remi/php73                                                                                                                         
   Loaded Configuration File               /etc/opt/remi/php73/php.ini              

                                    phalcon

                Web framework delivered as a C-extension for PHP

                     phalcon                 enabled
                Author            Phalcon Team and contributors
                Version           3.4.5
                Build Date        Oct 31 2019 08:35:22
                Powered by Zephir Version 0.10.16

                      Directive               Local Value Master Value
         phalcon.db.escape_identifiers        On          On
         phalcon.db.force_casting             Off         Off
         phalcon.orm.cast_on_hydrate          Off         Off
         phalcon.orm.column_renaming          On          On
         phalcon.orm.disable_assign_setters   Off         Off
         phalcon.orm.enable_implicit_joins    On          On
         phalcon.orm.enable_literals          On          On
         phalcon.orm.events                   On          On
         phalcon.orm.exception_on_failed_save Off         Off
         phalcon.orm.ignore_unknown_columns   Off         Off
         phalcon.orm.late_state_binding       Off         Off
         phalcon.orm.not_null_validations     On          On
         phalcon.orm.update_snapshot_on_save  On          On
         phalcon.orm.virtual_foreign_keys     On          On

**Reply by jwang628** [reply]
again command line works fine: /opt/remi/php73/root/usr/bin/php index.php

just index.php can not be called via apache :(

**Reply by jwang628** [reply]
just tried php56-php - the same error

looks the issue is with phalcon Version           3.4.5

**Comment by jwang628** [comment]
tried php56-php and the same error.

tried phalcon 4 and function 
$application->handle()->getContent()
Wrong number of parameter

tried the following without joy either:
$request = new Phalcon\Http\Request();
$uri=$request->getURI(); or $uri=$_SERVER['REQUEST_URI'];
$application->handle()->getContent($uri)

**Comment by Jeckerson** [comment]
You will need to enable logs to know where is the problem.

**Reply by jwang628** [reply]
how please?  nothing in /var/log/message nor in /var/log/httpd/

**Comment by Jeckerson** [comment]
As an alternative to console logs, you can configure Sentry in your project.

```
composer require sentry/sdk
```

And inside `index.php` at the begging of file:
```
Sentry\init(['dsn' => 'DSN_url' ]);
```

More docs here: https://docs.sentry.io/platforms/php/

It will give you possibility to see all errors of you application in one place.

**Reply by jwang628** [reply]
thanks a lot.  shall give it a try later.

**Reply by jwang628** [reply]
got error:

php index.php
PHP Fatal error:  Uncaught Error: Call to undefined function Sentry\init() in /var/www/html/jdm/public/index.php:2
Stack trace:
#0 {main}
  thrown in /var/www/html/jdm/public/index.php on line 2

**Reply by jwang628** [reply]
"composer require sentry/sdk" successful

 0 drwxrwxr-x. 12 james james   179 Mar 10 09:32 vendor
64 -rw-rw-r--.  1 james james 64108 Mar 10 09:32 composer.lock
 4 -rw-rw-r--.  1 james james    56 Mar 10 09:32 composer.json

**Reply by Jeckerson** [reply]
Call `Sentry\init()` after require `vendor/autoload.php` ...

**Reply by jwang628** [reply]
thanks a lot for the prompt response.

my index.php a few head lines:

<?php
require vendor/autoload.php;
Sentry\init(['dsn' => 'DSN_url' ]);
error_reporting(E_ALL);
try {
....

and new error:
php index.php
PHP Warning:  Use of undefined constant vendor - assumed 'vendor' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  Use of undefined constant autoload - assumed 'autoload' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  A non-numeric value encountered in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  A non-numeric value encountered in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  Division by zero in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  Use of undefined constant php - assumed 'php' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  require(NANphp): failed to open stream: No such file or directory in /var/www/html/jdm/public/index.php on line 2
PHP Fatal error:  require(): Failed opening required 'NANphp' (include_path='.:/usr/share/pear:/usr/share/php') in /var/www/html/jdm/public/index.php on line 2

**Reply by jwang628** [reply]
/opt/remi/php73/root/usr/bin/php index.php
PHP Warning:  Use of undefined constant vendor - assumed 'vendor' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  Use of undefined constant autoload - assumed 'autoload' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  A non-numeric value encountered in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  A non-numeric value encountered in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  Division by zero in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  Use of undefined constant php - assumed 'php' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2
PHP Warning:  require(NANphp): failed to open stream: No such file or directory in /var/www/html/jdm/public/index.php on line 2
PHP Fatal error:  require(): Failed opening required 'NANphp' (include_path='.:/opt/remi/php73/root/usr/share/pear:/opt/remi/php73/root/usr/share/php:/usr/share/pear:/usr/share/php') in /var/www/html/jdm/public/index.php on line 2

**Comment by jwang628** [comment]
my index.php a few head lines:

<?php
require vendor/autoload.php;
Sentry\init(['dsn' => 'DSN_url' ]);
error_reporting(E_ALL);
...

and new error: php index.php PHP Warning: Use of undefined constant vendor - assumed 'vendor' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2 PHP Warning: Use of undefined constant autoload - assumed 'autoload' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2 PHP Warning: A non-numeric value encountered in /var/www/html/jdm/public/index.php on line 2 PHP Warning: A non-numeric value encountered in /var/www/html/jdm/public/index.php on line 2 PHP Warning: Division by zero in /var/www/html/jdm/public/index.php on line 2 PHP Warning: Use of undefined constant php - assumed 'php' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2 PHP Warning: require(NANphp): failed to open stream: No such file or directory in /var/www/html/jdm/public/index.php on line 2 PHP Fatal error: require(): Failed opening required 'NANphp' (include_path='.:/usr/share/pear:/usr/share/php') in /var/www/html/jdm/public/index.php on line 2

**Comment by jwang628** [comment]
require vendor/autoload.php;
Sentry\init(['dsn' => 'DSN_url' ]);
error_reporting(E_ALL);

**Comment by jwang628** [comment]
index.php first 3 lines

require vendor/autoload.php;
Sentry\init(['dsn' => 'DSN_url' ]);
error_reporting(E_ALL);


and new error: php index.php PHP Warning: Use of undefined constant vendor - assumed 'vendor' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2 PHP Warning: Use of undefined constant autoload - assumed 'autoload' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2 PHP Warning: A non-numeric value encountered in /var/www/html/jdm/public/index.php on line 2 PHP Warning: A non-numeric value encountered in /var/www/html/jdm/public/index.php on line 2 PHP Warning: Division by zero in /var/www/html/jdm/public/index.php on line 2 PHP Warning: Use of undefined constant php - assumed 'php' (this will throw an Error in a future version of PHP) in /var/www/html/jdm/public/index.php on line 2 PHP Warning: require(NANphp): failed to open stream: No such file or directory in /var/www/html/jdm/public/index.php on line 2 PHP Fatal error: require(): Failed opening required 'NANphp' (include_path='.:/usr/share/pear:/usr/share/php') in /var/www/html/jdm/public/index.php on line 2

**Comment by jwang628** [comment]
/var/opt/remi/php73/log/php-fpm/www-error.log shows:

[13-Mar-2021 11:09:50 UTC] PHP Fatal error:  Uncaught Error: Class 'Security' not found in /var/www/html/jdm/public/index2.php:36
Stack trace:
#0 [internal function]: Closure->{closure}()
#1 [internal function]: Phalcon\Di\Service->resolve(NULL, Object(Phalcon\Di\FactoryDefault))
#2 [internal function]: Phalcon\Di->get('dispatcher', NULL)
#3 [internal function]: Phalcon\Di->getShared('dispatcher')
#4 /var/www/html/jdm/public/index2.php(164): Phalcon\Mvc\Application->handle()
#5 {main}
  thrown in /var/www/html/jdm/public/index2.php on line 36
  
the file does exist:
  james@testvm2 public]$ ll ../app/plugins/Security.php


I believe it is loaded as the same code worked in Windows and Centos6 :
error_reporting(E_ALL);
try {

        /**
         * Read the configuration
         */
        $config = new Phalcon\Config\Adapter\Ini(__DIR__ . '/../app/config/config.ini');

        $loader = new \Phalcon\Loader();

        /**
         * We're a registering a set of directories taken from the configuration file
         */
        $loader->registerDirs(
                array(
                        __DIR__ . $config->application->controllersDir,
                        __DIR__ . $config->application->pluginsDir,
                        __DIR__ . $config->application->libraryDir,
                        __DIR__ . $config->application->modelsDir,
                )
        )->register();

        /**
         * The FactoryDefault Dependency Injector automatically register the right services providing a full stack framework
         */
        $di = new \Phalcon\DI\FactoryDefault();

        /**
         * We register the events manager
         */
        $di->set('dispatcher', function() use ($di) {

                $eventsManager = $di->getShared('eventsManager');

                $security = new Security($di);

**Comment by Jeckerson** [comment]
Try to look at autoloading in Phalcon's demo projects, which uses composer PSR-4

Example:
https://github.com/phalcon/vokuro/blob/c768cc772738a12afe8f3c93594b37676bd0b447/composer.json#L47

* https://github.com/phalcon/vokuro
* https://github.com/phalcon/invo

**Reply by jwang628** [reply]
thanks a lot.  issue resolved.  folder had 700 and needs to be 755 so apache can read 

@Jeckerson : thanks indeed for your help.  best wishes and have a very good weekend ^
